contracts/YieldNotionalLever.sol

```diff
@@ -0,0 +1,510 @@
+            // Borrow to repay debt, move directly to the pool.
+            bytes memory data = bytes.concat(
+                bytes1(bytes1(uint8(uint256(Operation.CLOSE)))), // [0:1]
+                seriesId, // [1:7]
+                ilkId, // [7:13]
+                vaultId, // [13:25]
+                bytes32(ink), // [25:57]
+                bytes32(art) // [57:89]
+            );
+            bool success;
+            IlkInfo memory info = ilkInfo[ilkId];
+            IERC20 token = IERC20(info.join.asset());
+            success = info.join.flashLoan(
+                this, // Loan Receiver
+                address(token), // Loan Token
+                art, // Loan Amount: borrow exactly the debt to repay.
+                data
+            );
+
+            if (!success) revert FlashLoanFailure();
+
+            uint256 balance = token.balanceOf(address(this));
+            token.safeTransfer(msg.sender, balance);
+
+            emit Divested(
+                Operation.CLOSE,
+                vaultId,
+                seriesId,
+                msg.sender,
+                ink,
+                art
+            );
+        }
+        // Give the vault back to the sender, just in case there is anything left
+        giver.give(vaultId, msg.sender);
+    }
+
+    /// @notice Called by a flash lender, which can be `usdcJoin` or
+    ///     `daiJoin` or 'fyToken`. The primary purpose is to check conditions
+    ///     and route to the correct internal function.
+    ///
+    ///     This function reverts if not called through a flashloan initiated
+    ///     by this contract.
+    /// @param initiator The initator of the flash loan, must be `address(this)`.
+    /// @param borrowAmount The amount of fyTokens received.
+    /// @param fee The fee that is subtracted in addition to the borrowed
+    ///     amount when repaying.
+    /// @param data The data we encoded for the functions. Here, we only check
+    ///     the first byte for the router.
+    function onFlashLoan(
+        address initiator,
+        address token, // The token, not checked as we check the lender address.
+        uint256 borrowAmount,
+        uint256 fee,
+        bytes calldata data
+    ) external override returns (bytes32) {
+        Operation status = Operation(uint256(uint8(data[0])));
+        bytes6 seriesId = bytes6(data[1:7]);
+        bytes6 ilkId = bytes6(data[7:13]);
+        bytes12 vaultId = bytes12(data[13:25]);
+
+        // Test that the lender is either a fyToken contract or the join.
+        if (
+            msg.sender != address(IPool(ladle.pools(seriesId)).fyToken()) &&
+            msg.sender != address(ladle.joins(seriesId & ASSET_ID_MASK))
+        ) revert FlashLoanFailure();
+        // We trust the lender, so now we can check that we were the initiator.
+        if (initiator != address(this)) revert FlashLoanFailure();
+        // Decode the operation to execute and then call that function.
```

Comment:
`IERC20(token).approve(msg.sender, borrowAmount + fee + (status == Operation.CLOSE) ? borrowAmount : 0);`

Only if you want to be concise but obscure, your choice.